<a href="https://colab.research.google.com/github/pingao2019/DS-Unit-4-Sprint-2-Neural-Networks/blob/master/LS_DS_424_Hyperparameter_Tuning_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>

# Hyperparameter Tuning

## *Data Science Unit 4 Sprint 2 Assignment 4*

## Your Mission, should you choose to accept it...

To hyperparameter tune and extract every ounce of accuracy out of this telecom customer churn dataset: [Available Here](https://lambdaschool-data-science.s3.amazonaws.com/telco-churn/WA_Fn-UseC_-Telco-Customer-Churn+(1).csv)

## Requirements

- Load the data
- Clean the data if necessary (it will be)
- Create and fit a baseline Keras MLP model to the data.
- Hyperparameter tune (at least) the following parameters:
 - batch_size
 - training epochs
 - optimizer
 - learning rate (if applicable to optimizer)
 - momentum (if applicable to optimizer)
 - activation functions
 - network weight initialization
 - dropout regularization
 - number of neurons in the hidden layer
 
 You must use Grid Search and Cross Validation for your initial pass of the above hyperparameters
 
 Try and get the maximum accuracy possible out of this data! You'll save big telecoms millions! Doesn't that sound great?


In [6]:
##### Your Code Here #####
import pandas as pd
url = 'https://lambdaschool-data-science.s3.amazonaws.com/telco-churn/WA_Fn-UseC_-Telco-Customer-Churn+(1).csv'
df = pd.read_csv(url)
df

customerID  gender  SeniorCitizen  ... MonthlyCharges TotalCharges  Churn
0     7590-VHVEG  Female              0  ...          29.85        29.85     No
1     5575-GNVDE    Male              0  ...          56.95       1889.5     No
2     3668-QPYBK    Male              0  ...          53.85       108.15    Yes
3     7795-CFOCW    Male              0  ...          42.30      1840.75     No
4     9237-HQITU  Female              0  ...          70.70       151.65    Yes
...          ...     ...            ...  ...            ...          ...    ...
7038  6840-RESVB    Male              0  ...          84.80       1990.5     No
7039  2234-XADUH  Female              0  ...         103.20       7362.9     No
7040  4801-JZAZL  Female              0  ...          29.60       346.45     No
7041  8361-LTMKD    Male              1  ...          74.40        306.6    Yes
7042  3186-AJIEK    Male              0  ...         105.65       6844.5     No

[7043 rows x 21 columns]

In [3]:
df.shape

(7043, 21)

Data clean

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [8]:
df.Churn.value_counts(normalize=True)

No     0.73463
Yes    0.26537
Name: Churn, dtype: float64

In [0]:
# Removing unwanted coloumns
df.drop(['customerID'],axis=1,inplace=True)

In [0]:
df.isnull().sum()

In [0]:
# set X features matrix and y target vector
features = list(df)[:-1]
target = list(df)[-1]
X = df[features]
y = df[target].apply(lambda x: 1 if x == 'Yes' else 0)

In [14]:
# Target encode all categories
!pip install category_encoders

     |████████████████████████████████| 81kB 2.4MB/s 


In [16]:

from category_encoders.target_encoder import TargetEncoder

encoder = TargetEncoder()
X_encoded = encoder.fit_transform(X, y)
encoder.get_feature_names

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


<bound method TargetEncoder.get_feature_names of TargetEncoder(cols=['gender', 'Partner', 'Dependents', 'PhoneService',
                    'MultipleLines', 'InternetService', 'OnlineSecurity',
                    'OnlineBackup', 'DeviceProtection', 'TechSupport',
                    'StreamingTV', 'StreamingMovies', 'Contract',
                    'PaperlessBilling', 'PaymentMethod', 'TotalCharges'],
              drop_invariant=False, handle_missing='value',
              handle_unknown='value', min_samples_leaf=1, return_df=True,
              smoothing=1.0, verbose=0)>

In [0]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [20]:
# Normalized by Standard scale

scaler = StandardScaler()
X_norm = scaler.fit_transform(X_encoded)
X_norm

array([[ 1.00955867, -0.43991649, -1.03453023, ...,  1.39974289,
        -1.16032292, -0.15193369],
       [-0.99053183, -0.43991649,  0.96662231, ..., -0.55473949,
        -0.25962894, -0.15193369],
       [-0.99053183, -0.43991649,  0.96662231, ..., -0.55473949,
        -0.36266036, -0.15193369],
       ...,
       [ 1.00955867, -0.43991649, -1.03453023, ...,  1.39974289,
        -1.1686319 , -0.15193369],
       [-0.99053183,  2.27315869, -1.03453023, ..., -0.55473949,
         0.32033821, -0.15193369],
       [-0.99053183, -0.43991649,  0.96662231, ..., -0.73368684,
         1.35896134, -0.15193369]])

In [21]:
X_norm.shape, X_norm[0].shape

((7043, 19), (19,))

In [31]:
# Baseline keras ML model
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Flatten, Dense

model = Sequential()
model.add(Flatten(input_shape=X_norm[0].shape))
model.add(Dense(10))
model.add(Dense(1, activation='sigmoid'))


model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

baseline = model.fit(X_norm, y, validation_split=0.2, epochs=20)

Epoch 1/20
177/177 [==============================] - 1s 3ms/step - loss: 0.6090 - accuracy: 0.6858 - val_loss: 0.4798 - val_accuracy: 0.7700
Epoch 2/20
177/177 [==============================] - 1s 3ms/step - loss: 0.4347 - accuracy: 0.7920 - val_loss: 0.4156 - val_accuracy: 0.8084
Epoch 3/20
177/177 [==============================] - 1s 3ms/step - loss: 0.4019 - accuracy: 0.8136 - val_loss: 0.4068 - val_accuracy: 0.8148
Epoch 4/20
177/177 [==============================] - 0s 3ms/step - loss: 0.3962 - accuracy: 0.8175 - val_loss: 0.4047 - val_accuracy: 0.8197
Epoch 5/20
177/177 [==============================] - 0s 3ms/step - loss: 0.3939 - accuracy: 0.8200 - val_loss: 0.4049 - val_accuracy: 0.8183
Epoch 6/20
177/177 [==============================] - 1s 3ms/step - loss: 0.3931 - accuracy: 0.8184 - val_loss: 0.4055 - val_accuracy: 0.8126
Epoch 7/20
177/177 [==============================] - 1s 3ms/step - loss: 0.3930 - accuracy: 0.8214 - val_loss: 0.4052 - val_accuracy: 0.8148
Epoch 

In [32]:


model.summary()

Model: "sequential_7"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_5 (Flatten)          (None, 19)                0         
_________________________________________________________________
dense_14 (Dense)             (None, 10)                200       
_________________________________________________________________
dense_15 (Dense)             (None, 1)                 11        
Total params: 211
Trainable params: 211
Non-trainable params: 0
_________________________________________________________________


In [33]:
from tensorflow.keras.callbacks import EarlyStopping

stop = EarlyStopping(monitor='val_accuracy', min_delta=.001, patience=2)
model.fit(X_norm, y, 
          validation_split=0.2, 
          epochs=100, 
          batch_size=None, 
          callbacks=[stop])

Epoch 1/100
177/177 [==============================] - 0s 2ms/step - loss: 0.3917 - accuracy: 0.8220 - val_loss: 0.4041 - val_accuracy: 0.8105
Epoch 2/100
177/177 [==============================] - 0s 2ms/step - loss: 0.3917 - accuracy: 0.8204 - val_loss: 0.4025 - val_accuracy: 0.8091
Epoch 3/100
177/177 [==============================] - 0s 2ms/step - loss: 0.3917 - accuracy: 0.8222 - val_loss: 0.4033 - val_accuracy: 0.8119
Epoch 4/100
177/177 [==============================] - 0s 2ms/step - loss: 0.3918 - accuracy: 0.8232 - val_loss: 0.4038 - val_accuracy: 0.8105
Epoch 5/100
177/177 [==============================] - 0s 2ms/step - loss: 0.3912 - accuracy: 0.8193 - val_loss: 0.4041 - val_accuracy: 0.8098


In [0]:
# Hyper param tuning
from sklearn.model_selection import GridSearchCV
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier

# We use an api wrapper to that we can work with scikit_learn's gridsearchCV
# KerasClassifier's build_fn param needs a function of the model builder to 
# iterate over. So let's make that function
def create_model():
  model = Sequential()
  model.add(Flatten(input_shape=X_norm[0].shape))
  model.add(Dense(10, activation='relu'))
  model.add(Dense(1, activation='sigmoid'))

  model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
  return model

# now instantiate KerasClassifier with your function, first time calling model
model = KerasClassifier(build_fn=create_model, verbose=1)

In [37]:
#  grid search some parameters
param_grid = {'batch_size': [16,32,64,128],
              'epochs': [20]}

# Create Grid Search
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=-3)
grid_result = grid.fit(X_norm, y)

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}")

Epoch 1/20
353/353 [==============================] - 1s 2ms/step - loss: 0.5301 - accuracy: 0.7332
Epoch 2/20
353/353 [==============================] - 1s 2ms/step - loss: 0.4193 - accuracy: 0.8026
Epoch 3/20
353/353 [==============================] - 1s 2ms/step - loss: 0.4097 - accuracy: 0.8110
Epoch 4/20
353/353 [==============================] - 1s 2ms/step - loss: 0.4057 - accuracy: 0.8117
Epoch 5/20
353/353 [==============================] - 1s 2ms/step - loss: 0.4029 - accuracy: 0.8136
Epoch 6/20
353/353 [==============================] - 1s 1ms/step - loss: 0.4008 - accuracy: 0.8151
Epoch 7/20
353/353 [==============================] - 1s 2ms/step - loss: 0.3993 - accuracy: 0.8175
Epoch 8/20
353/353 [==============================] - 1s 2ms/step - loss: 0.3980 - accuracy: 0.8181
Epoch 9/20
353/353 [==============================] - 1s 2ms/step - loss: 0.3963 - accuracy: 0.8166
Epoch 10/20
353/353 [==============================] - 1s 2ms/step - loss: 0.3957 - accuracy: 0.8213

from above, Best: 0.8182591438293457 using {'batch_size': 32, 'epochs': 20

In [38]:
# learning rate as a hyperparam, specify it in the create_model

from tensorflow.keras.optimizers import Nadam

def create_model(learning_rate=0.01):
  model = Sequential()
  model.add(Flatten(input_shape=X_norm[0].shape))
  model.add(Dense(23, activation='relu'))
  model.add(Dense(1, activation='sigmoid'))

  optimizer = Nadam(learning_rate=learning_rate)
  model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
  return model

# create model
model = KerasClassifier(build_fn=create_model, verbose=0)


param_grid = {'learning_rate': [.001, .01, .1, .2, .3],
              'epochs': [30]}

# Create Grid Search
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=-3)
grid_result = grid.fit(X_norm, y)

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}") 

Best: 0.8155616879463196 using {'epochs': 30, 'learning_rate': 0.001}
Means: 0.8155616879463196, Stdev: 0.0037869536495924597 with: {'epochs': 30, 'learning_rate': 0.001}
Means: 0.8105922341346741, Stdev: 0.006765183836640906 with: {'epochs': 30, 'learning_rate': 0.01}
Means: 0.7685731768608093, Stdev: 0.02751998696708586 with: {'epochs': 30, 'learning_rate': 0.1}
Means: 0.7797789454460144, Stdev: 0.0338879512560142 with: {'epochs': 30, 'learning_rate': 0.2}
Means: 0.7343446493148804, Stdev: 0.004863621649161728 with: {'epochs': 30, 'learning_rate': 0.3}
Means: 0.7346286535263061, Stdev: 0.004757632470334884 with: {'epochs': 30, 'learning_rate': 0.5}


Above result:  Best: 0.8155616879463196 using {'epochs': 30, 'learning_rate': 0.001}

In [41]:
# continue to select best optimizer

def create_model(optimizer='adam', init='uniform' ):
  model = Sequential()
  model.add(Flatten(input_shape=X_norm[0].shape))
  model.add(Dense(23, activation='relu'))
  model.add(Dense(1, activation='sigmoid'))

  #optimizer = Nadam(learning_rate=learning_rate)
  model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
  return model

# create model
model = KerasClassifier(build_fn=create_model, verbose=0)

param_grid = {'epochs': [40],
              'batch_size': [32],
              'optimizer': ['rmsprop', 'adam', 'nadam']
              }

# Create Grid Search
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=-3)
grid_result = grid.fit(X_norm, y)

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}") 

Best: 0.8191107988357544 using {'batch_size': 32, 'epochs': 40, 'optimizer': 'nadam'}
Means: 0.816413152217865, Stdev: 0.004068224293000359 with: {'batch_size': 32, 'epochs': 40, 'optimizer': 'rmsprop'}
Means: 0.8169814229011536, Stdev: 0.005740121006222151 with: {'batch_size': 32, 'epochs': 40, 'optimizer': 'adam'}
Means: 0.8191107988357544, Stdev: 0.0034209116453548676 with: {'batch_size': 32, 'epochs': 40, 'optimizer': 'nadam'}


so the result Best: 0.8191107988357544 using {'batch_size': 32, 'epochs': 40, 'optimizer': 'nadam'}

In [43]:
# letus try the best parameters, 'batch_size': 32, 'epochs': 40, 'optimizer': 'nadam','learning_rate': 0.001

def create_model(learning_rate=0.001):
  model = Sequential()
  model.add(Flatten(input_shape=X_norm[0].shape))
  model.add(Dense(10,activation='relu'))
  model.add(Dense(1, activation='sigmoid'))

  optimizer = Nadam(learning_rate=learning_rate)
  model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
  return model

# create model
model = KerasClassifier(build_fn=create_model, verbose=0)

param_grid = {
              'epochs': [80],
              'batch_size': [32],
              #'optimizer': ['nadam']
              }

# Create Grid Search
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=-3)
grid_result = grid.fit(X_norm, y)

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}")


Best: 0.815561580657959 using {'batch_size': 32, 'epochs': 80}
Means: 0.815561580657959, Stdev: 0.004473356939781269 with: {'batch_size': 32, 'epochs': 80}


In [50]:
# apply early stopping
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.layers import Dropout

stop = EarlyStopping(monitor='val_accuracy', min_delta=0.001, patience=20)

model = Sequential()
model.add(Flatten(input_shape=X_norm[0].shape))
model.add(Dense(20, activation='relu'))
Dropout(.1)
model.add(Dense(20, activation='relu'))
Dropout(.025),
model.add(Dense(1, activation='sigmoid'))
optimizer= Nadam(learning_rate=0.001)

model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])

model.fit(X_norm, y, validation_split=0.20, epochs=100, callbacks=[stop])


print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")

Epoch 1/100
177/177 [==============================] - 0s 2ms/step - loss: 0.4984 - accuracy: 0.7352 - val_loss: 0.4315 - val_accuracy: 0.7913
Epoch 2/100
177/177 [==============================] - 0s 2ms/step - loss: 0.4135 - accuracy: 0.8019 - val_loss: 0.4166 - val_accuracy: 0.8027
Epoch 3/100
177/177 [==============================] - 0s 2ms/step - loss: 0.4009 - accuracy: 0.8111 - val_loss: 0.4115 - val_accuracy: 0.8091
Epoch 4/100
177/177 [==============================] - 0s 2ms/step - loss: 0.3947 - accuracy: 0.8147 - val_loss: 0.4103 - val_accuracy: 0.8041
Epoch 5/100
177/177 [==============================] - 0s 2ms/step - loss: 0.3916 - accuracy: 0.8170 - val_loss: 0.4071 - val_accuracy: 0.8155
Epoch 6/100
177/177 [==============================] - 0s 2ms/step - loss: 0.3892 - accuracy: 0.8193 - val_loss: 0.4067 - val_accuracy: 0.8141
Epoch 7/100
177/177 [==============================] - 0s 2ms/step - loss: 0.3875 - accuracy: 0.8213 - val_loss: 0.4064 - val_accuracy: 0.8126

Above results: Best: 0.815561580657959 using {'batch_size': 32, 'epochs': 80}

## Stretch Goals:

- Try to implement Random Search Hyperparameter Tuning on this dataset
- Try to implement Bayesian Optimiation tuning on this dataset using hyperas or hyperopt (if you're brave)
- Practice hyperparameter tuning other datasets that we have looked at. How high can you get MNIST? Above 99%?
- Study for the Sprint Challenge
 - Can you implement both perceptron and MLP models from scratch with forward and backpropagation?
 - Can you implement both perceptron and MLP models in keras and tune their hyperparameters with cross validation?